### ISSUE TITLE
JuliaCall fails with segmentation fault (core dumped) 

### ISSUE THREAD

**Issue by jalcam** [JuliaInterop/JuliaCall#238]
I have Julia installed with Juliaup (1.10 and 1.11). When I try to setup juliacall in R with 1.10 I get a segfault:

```r
> library(JuliaCall)
> julia_setup(JULIA_HOME = "/home/juanjose/.julia/juliaup/julia-1.10.6+0.x64.linux.gnu/bin")
Julia version 1.10.6 at location /home/juanjose/.julia/juliaup/julia-1.10.6+0.x64.linux.gnu/bin will be used.
Loading setup script for JuliaCall...
Segmentation fault (core dumped)
```

 when I try with Julia 1.11 the error is:

```r
> library(JuliaCall)
> julia_setup(JULIA_HOME = "/home/juanjose/.julia/juliaup/julia-1.11.1+0.x64.linux.gnu/bin")
Julia version 1.11.1 at location /home/juanjose/.julia/juliaup/julia-1.11.1+0.x64.linux.gnu/bin will be used.
Error in juliacall_initialize(.julia$dll_file, .julia$bin_dir, img_abs_path) : 
  jl_stdout_obj - /home/juanjose/.julia/juliaup/julia-1.11.1+0.x64.linux.gnu/bin/../lib/libjulia.so.1.11: undefined symbol: jl_stdout_obj
```

In Julia I've tried:

```julia
julia> using RCall

julia> RCall.Rhome
"/usr/lib/R"

julia> RCall.libR
"/usr/lib/R/lib/libR.so"

julia> RCall.validate_libR(RCall.libR)
true

julia> ENV["R_HOME"] = "/usr/lib/R"
"/usr/lib/R"

julia> Pkg.build("RCall")
    Building Conda → `~/.julia/scratchspaces/44cfe95a-1eb2-52ea-b672-e2afdf69b78f/b19db3927f0db4151cb86d073689f2428e524576/build.log`
    Building RCall → `~/.julia/scratchspaces/44cfe95a-1eb2-52ea-b672-e2afdf69b78f/db17ec90d9f904b79e7877a764fdf95ff5c5f315/build.log`
```

But the segfault persists when trying JuliaCall

**Comment by ViralBShah** [comment]
Do you have something in your LD_LIBRARY_PATH or DYLD_LIBRARY_PATH or LD_PRELOAD?

**Comment by jalcam** [comment]
> Do you have something in your LD_LIBRARY_PATH or DYLD_LIBRARY_PATH or LD_PRELOAD?

I have this:

```
> echo $LD_LIBRARY_PATH
/opt/openfoam12/thirdparty/platforms/linux64Gcc/gperftools-svn/lib:/usr/lib/x86_64-linux-gnu:/opt/openfoam12/platforms/linux64GccDPInt32Opt/lib/paraview-5.11:/opt/openfoam12/platforms/linux64GccDPInt32Opt/lib/openmpi-system:/opt/openfoam12/thirdparty/platforms/linux64GccDPInt32/lib/openmpi-system:/usr/lib/x86_64-linux-gnu/openmpi/lib:/home/jalcam/OpenFOAM/jalcam-12/platforms/linux64GccDPInt32Opt/lib:/opt/site/12/platforms/linux64GccDPInt32Opt/lib:/opt/openfoam12/platforms/linux64GccDPInt32Opt/lib:/opt/openfoam12/thirdparty/platforms/linux64GccDPInt32/lib:/opt/openfoam12/platforms/linux64GccDPInt32Opt/lib/dummy

> echo $DYLD_LIBRARY_PATH

> echo $LD_PRELOAD 

```

I use OpenFoam in the same computer.

**Comment by ViralBShah** [comment]
Can you unset that variable and see if it still crashes?

**Comment by jalcam** [comment]
> Can you unset that variable and see if it still crashes?

I have unset the variable now:

```
> echo $LD_LIBRARY_PATH


```

But I'm getting the same error:

```
> library(JuliaCall)
> julia_setup(JULIA_HOME = "/home/juanjose/.julia/juliaup/julia-1.10.6+0.x64.linux.gnu/bin")
Julia version 1.10.6 at location /home/juanjose/.julia/juliaup/julia-1.10.6+0.x64.linux.gnu/bin will be used.
Loading setup script for JuliaCall...
Segmentation fault (core dumped)
```

**Comment by marc-milgrom-ncl** [comment]
I have downloaded Chris's latest patch branch, and devtools:check seg faults on 1.11. When I override it to build with JULIA_HOME set to the latest 1.9 release, it builds, but then julia_setup causes R to seg fault.
I'm running R 4.4.2.

**Comment by marc-milgrom-ncl** [comment]
I think I'm making progress here. julia_setup is failing, and I see this:
`julia> using Libdl

julia> Libdl.dlopen("/usr/local/lib/R/library/methods/libs/methods.so")
ERROR: could not load library "/usr/local/lib/R/library/methods/libs/methods.so"
libR.so: cannot open shared object file: No such file or directory
Stacktrace:
 [1] dlopen(s::String, flags::UInt32; throw_error::Bool)
   @ Base.Libc.Libdl ./libdl.jl:117
 [2] dlopen
   @ ./libdl.jl:116 [inlined]
 [3] dlopen(s::String)
   @ Base.Libc.Libdl ./libdl.jl:116
 [4] top-level scope
   @ REPL[2]:1`
The `methods` package is installed, but I had a similar error with `libRblas.so` until I rebuilt R with the `with-blas` option. It seems like JuliaCall is expecting all the R libraries to be statically linked into libR.so or something? It's not possible to build R that way to my knowledge.

**Comment by ViralBShah** [comment]
How was the R you are using built and installed? This may be an issue with some LD_LIBRARY_PATH type setting.

I always wanted R in Yggdrasil, but IIRC, it does not support cross compilation. But maybe we can just package the upstream binaries like we do for MKL. And at least that way, we can have something that works reliably.

**Comment by marc-milgrom-ncl** [comment]
I don't have any LD_LIBRARY_PATH set, but I tried setting `ENV["LD_LIBRARY_PATH"]="/usr/local/lib:/usr/local/lib/R/library"` in Julia before running `setup.jl` to no avail.
I built R from the current 4.4.2 release source with:
```
sudo ./configure --enable-R-shlib --enable-memory-profiling  --with-blas --prefix=/usr/local
```
and then `sudo make` and `sudo make install`.

To be clear I only built it from source to enable debugging of the segmentation fault I was having with the packaged R binaries, which are stripped. This didn't help as `R -d gdb` isn't catching the SEGV, I think because it's actually in Julia, not R, but that's a guess.

**Comment by ViralBShah** [comment]
Is `methods.so` coming from JuliaCall or from R? The issue with `methods.so` seems to do with resolving all the dependencies for the shared library.

**Comment by marc-milgrom-ncl** [comment]
The `methods.so` library comes from `methods`, one of the expected R packages called by RCall from Julia. It happens to be one of the recommended packages included in the standard R build, i.e. unless `-without-recommended-packages` is specified to `configure`.

**Comment by marc-milgrom-ncl** [comment]
I was about to test if building R without the recommended packages and then installing them after build and install would change the behavior, but it's now 17:30 here so I'll pick this up tomorrow.

**Comment by ViralBShah** [comment]
Is there an issue with 1.10 too?

**Comment by marc-milgrom-ncl** [comment]
It also SEGV'd the last time I tried it, but that might have been with the packaged R binaries so not necessarily the same issue.

**Comment by simonbeyer1** [comment]
I had the exact same problem yesterday. Downgrading to Ubuntu 22.04.5 LTS (Jammy Jellyfish) solved the issue for me.

**Comment by marc-milgrom-ncl** [comment]
Thanks. I'll spin up a 22 VM and see if that works.

**Comment by ViralBShah** [comment]
How about with Julia 1.10?

**Comment by marc-milgrom-ncl** [comment]
check() runs ok with 1.10, and the package seems to function.
So there seem to be issues with both Ubuntu 24 and Julia 1.11.

**Comment by jalcam** [comment]
I have installed R on Ubuntu 24.04 using binaries via https://eddelbuettel.github.io/r2u/. Could it have something to do with it? I think it should work with Julia 1.10 anyway but the segfault persists

**Comment by marc-milgrom-ncl** [comment]
As far I can I, and the commenter above, can tell, Ubuntu 24 is incompatible with JuliaCall. It will not build with julia 1.10 or 1.11, and while it seemed to build with 1.9, it segfaulted as soon as I attempted to run it.

**Comment by edwardlavender** [comment]
I experienced the same issue on Ubuntu 24.04.1 LTS with Julia 1.11 and Julia 1.10.5. I haven't tried downgrading Ubuntu version.

```
# > library(JuliaCall)
# > julia_setup(rebuild = TRUE, force = TRUE)
# Julia version 1.10.5 at location /home/edward/.julia/juliaup/julia-1.10.5+0.x64.linux.gnu/bin will be used.
# Loading setup script for JuliaCall...
# Segmentation fault (core dumped)
```

**Comment by trafficfan** [comment]
```
> library(JuliaCall)
> julia_setup()
Julia version 1.11.2 at location C:\Users\admin\.julia\juliaup\julia-1.11.2+0.x64.w64.mingw32\bin will be used.
Loading setup script for JuliaCall...
错误于.julia$cmd(paste0(Rhomeset, "Base.include(Main,\"", system.file("julia/setup.jl", : 
  Error happens when you try to execute command ENV["R_HOME"] = "C:/PROGRA~1/R/R-devel";Base.include(Main,"C:/Users/admin/AppData/Local/R/win-library/4.5/JuliaCall/julia/setup.jl") in Julia.
                        To have more helpful error messages,
                        you could considering running the command in Julia directly
```
I meet the same issue on win11,could someone help?

**Comment by mkitti** [comment]
> > library(JuliaCall)\n> julia_setup()\nJulia version 1.11.2 at location C:\\Users\\admin\\.julia\\juliaup\\julia-1.11.2+0.x64.w64.mingw32\\bin will be used.\nLoading setup script for JuliaCall...\n错误于.julia$cmd(paste0(Rhomeset, \"Base.include(Main,\\\"\", system.file(\"julia/setup.jl\", : \n  Error happens when you try to execute command ENV[\"R_HOME\"] = \"C:/PROGRA~1/R/R-devel\";Base.include(Main,\"C:/Users/admin/AppData/Local/R/win-library/4.5/JuliaCall/julia/setup.jl\") in Julia.\n                        To have more helpful error messages,\n                        you could considering running the command in Julia directly\nI meet the same issue on win11,could someone help?

This does not seem like the same error. Could you create another issue?

**Comment by keithmcnulty** [comment]
Just upgraded my EC2 to Ubuntu 24 today and and seeing the same issue with JuliaCall with Julia 1.11, 1.10 - also went back to 1.6.  Does indeed seem to be an Ubuntu 24 issue.

**Comment by mkitti** [comment]
How are you installing R?

**Comment by keithmcnulty** [comment]
> How are you installing R?

`sudo apt install --no-install-recommends r-base` latest version (4.4.2) from CRAN's repositiories.

I have reinstalled everything on Ubuntu 22 and all works smoothly, so does seem to be some issue with Ubuntu 24.

**Comment by mkitti** [comment]
I wonder if it would make sense to install R via conda-forge or other means. Otherwise, we may need to prepend Julia's libraries to the `LD_LIBRARY_PATH`.

**Comment by mlechu** [comment]
This seems to be an `LD_LIBRARY_PATH` issue combined with a bad version of libunwind.  In particular:

- Running R as packaged from apt uses a wrapper shell script (`/usr/bin/R`) which starts the real R binary (`/usr/lib/R/bin/exec/R`) after setting a bunch of environment variables, notably adding `/usr/lib/x86_64-linux-gnu` to `LD_LIBRARY_PATH` for some reason.  This is why changing `LD_LIBRARY_PATH` before starting R is ineffective.
- Julia then loads `libunwind.so.8` from `/usr/lib/x86_64-linux-gnu/` instead of from julia (e.g.  `/home/user/.julia/juliaup/julia-1.11.3+0.x64.linux.gnu/bin/../lib/julia/`).  `apt info libunwind8` shows `Version: 1.6.2-3build1` about the system package.
- Julia does prepare for segfaults from libunwind, but when I load libunwind 1.6.2, it uses `sigprocmask` to make the segfault handler ineffective (right before segfaulting).  More investigation is needed into what the actual bug is, but 1.6.2 also gives us a segmentation fault in plain Julia (MWE: `julia -e "foo(n) = n > 0 ? foo(n-1) : [1,2,3][4]; foo(10)"`, thanks @topolarity!) Also, ubuntu 22's default libunwind8 (1.3.2-2build2.1) doesn't have this issue.

The environment variables used are the ones set when R starts, so JuliaCall can't fix this by unsetting or prepending to `LD_LIBRARY_PATH`.  (Admittedly I did try.)

### Workaround
If you need to get JuliaCall working now, you can either preload correct libunwind, e.g.
```bash
LD_PRELOAD="/home/user/.julia/juliaup/julia-1.11.3+0.x64.linux.gnu/bin/../lib/julia/libunwind.so.8" R
```
or write a new R wrapper script similar to `/usr/bin/R` that doesn't set `LD_LIBRARY_PATH`.
```bash
#!/bin/bash
# subset of the shell wrapper for R executable.

R_HOME=/usr/lib/R
export R_HOME
R_SHARE_DIR=/usr/share/R/share
export R_SHARE_DIR
R_INCLUDE_DIR=/usr/share/R/include
export R_INCLUDE_DIR
R_DOC_DIR=/usr/share/R/doc
export R_DOC_DIR
exec /usr/lib/R/bin/exec/R
```

**Comment by topolarity** [comment]
> edit: Possibly be related to https://github.com/libunwind/libunwind/issues/637

Yeah, confirmed this - the issue bisects to exactly the same change

**Comment by nalimilan** [comment]
Good catch! So to me it seems that the R wrapper scripts used by some distributions are bad citizens as they set `LD_LIBRARY_PATH`, which is generally a recipe for problems. FWIW Fedora doesn't seem to do this so there are other solutions. Maybe we should report this to Ubuntu?

**Comment by jalcam** [comment]
I have installed R on Ubuntu 24.04 using binaries from [this source](https://eddelbuettel.github.io/r2u/). These binaries are not maintained by Ubuntu. My `LD_LIBRARY_PATH` is empty.

**Comment by mkitti** [comment]
@jalcam , have you tried setting LD_PRELOAD as in https://github.com/JuliaInterop/JuliaCall/issues/238#issuecomment-2620093298 ?

**Comment by katrinabrock** [comment]
Same issue...LD_PRELOAD workaround worked for me. Thanks @mlechu

**Comment by katrinabrock** [comment]
It also works to load Julia's libuwind after starting R but before running `julia_setup` like this:
```
my_version <- '1.11.4+0.x64.linux.gnu'
dyn.load(paste0('~/.julia/juliaup/julia-', my_version, '/lib/julia/libunwind.so.8'))
```

**Comment by ViralBShah** [comment]
@staticfloat Is the same kind of fix you applied to libsuitesparseconfig applicable to Julia itself so that we load the correct libunwind (and not get influenced by the LD_LIBRARY_PATH etc)?

**Comment by topolarity** [comment]
> [@staticfloat](https://github.com/staticfloat) Is the same kind of fix you applied to libsuitesparseconfig applicable to Julia itself so that we load the correct libunwind (and not get influenced by the LD_LIBRARY_PATH etc)?

It's slightly more work than the suitesparse case - The fix here is to make sure the stdlib JLL's have absolute paths: https://github.com/JuliaLang/julia/issues/55478
